Изменения, которые будут включены в коммит: валидации и скоупы

RamzanBersanov opened the pull request.

RamzanBersanov committed:
Изменения, которые будут включены в коммит: валидации и скоупы

VidgarVii commented on app/models/category.rb:
```diff
@@ -1,3 +1,7 @@
+
+  validates :title, presence: true
+
+  default_scope { order(title: :asc) }
```
`default_scope` глобально меняет поведение всех запросов к модели
поэтому его [рекомендуют](https://github.com/rubocop-hq/rails-style-guide#macro-style-methods) записывать в самом верху класса

VidgarVii commented on app/models/question.rb:
```diff
@@ -2,4 +2,16 @@ class Question < ApplicationRecord
+  validate :validate_answer_number
+
+  private
+  def validate_answer_number
```
валидация в модели Вопрос не будет препятствовать созданию новых ответов в количестве больше 4, потому что она не будет вызываться при добавлении новых ответов к вопросу

валидации в классе Вопрос вызываются при создании или изменении объектов класса Вопрос
при создании объектов класса Ответ валидации в классе Вопрос не вызываются

видимо данная валидации не тестировалась

VidgarVii commented on app/models/question.rb:
```diff
@@ -2,4 +2,16 @@ class Question < ApplicationRecord
+  private
+  def validate_answer_number
+    if answers.present?
+      unless answers.count.between?(1,4)
```
с такой проверкой есть следующие ньюансы:
при наличие уже 4-х объектов в базе можно создать 5-ый
и ошибку это не вызовет, что противоречит ТЗ
в отсутствии ответов, т.е. `count = 0` мы наоборот получим ошибку и первый ответ создать не сможем

VidgarVii commented on app/models/test.rb:
```diff
@@ -6,10 +6,16 @@ class Test < ApplicationRecord
+  scope :category_title, -> (category_title) { joins(:category)
+                                                .where(categories: {title: category_title})
+                                                .order(title: :desc)
+                                                .pluck(:title) }
```
у скоупов есть важная особенность - они должны возвращать chainable объекты (объекты `ActiveRecord::Relation`)
благодаря этому с их использованием можно строить цепочки вызовов
скоупы можно чейнить с квери-методами  `ActiveRecord`, к примеру `where(level: ...).order(id: :desc).created_last_week ...`, с методами ассоциаций
квери-методы также возвращают объекты `ActiveRecord::Relation`
НО!
`pluck` вернет обычный массив, что нарушит эту конвенцию
поэтому, если скоуп должен вернуть объект или массив, но не `ActiveRecord::Relation`, то его лучше реализовывать обычными методами класса

здесь можно создать скоуп, который будет делать, к примеру, все кроме `pluck`
а при желании сохранить прежнюю функциональность можно оставить плюс к скоупу метод класса и уже в методе класса (из прошлого задания) убрать все запросы, вызвать этот скоуп и зачейнить его "плаком"

VidgarVii commented on app/models/test.rb:
```diff
@@ -6,10 +6,16 @@ class Test < ApplicationRecord
-      .order(title: :desc)
-      .pluck(:title)
-  end
+  validates :title, presence: true, uniqueness: true
```
ТЗ: "Может существовать только один Тест с данным названием и уровнем"
здесь проверка на уникальность названия без связи с уровнем
нужна же проверка, что с таким названием только один тест именно такого уровня (можно иметь тест с таким же названием, но другого уровня)

VidgarVii commented on app/models/test.rb:
```diff
@@ -6,10 +6,16 @@ class Test < ApplicationRecord
-      .pluck(:title)
-  end
+  validates :title, presence: true, uniqueness: true
+  validates :level, :numericality => { only_integer: true, greater_than: 0 }, uniqueness: true
```
uniqueness тут лишний

RamzanBersanov committed:
/# Изменения, которые будут включены в коммит: исправлена валидация теста и ответа

VidgarVii commented on app/models/test.rb:
```diff
@@ -6,10 +6,18 @@ class Test < ApplicationRecord
-      .where(categories: {title: category_title})
-      .order(title: :desc)
-      .pluck(:title)
+  validates :title, uniqueness: { scope: :level, message: "can only be one test with such title and level" }
```
[в документации](https://guides.rubyonrails.org/active_record_validations.html#uniqueness) к хелперу `uniqueness` сказано следующее:
> It does not create a uniqueness constraint in the database, so it may happen that two different database connections create two records with the same value for a column that you intend to be unique. To avoid that, you must create a unique index on that column in your database.

другими словами, чтобы эта была настоящая уникальность (не только на уровне приложения, но и на уровне базы), нужно в базе создать уникальный индекс

про правильную миграцию можно посмотреть к примеру на [СО](https://stackoverflow.com/questions/32927262/create-migration-with-uniqueness-to-a-scope)

примерно таким образом:

```ruby
class AddIndexToTests < ActiveRecord::Migration[5.2]
  def change
    add_index :tests, [:title, :level], unique: true
  end
end
```

RamzanBersanov committed:
Изменения, которые будут включены в коммит:
	новый файл:    db/migrate/20221118085741_add_index_to_tests.rb
	изменено:      db/schema.rb
